group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/CurrentAssignmentBuilder.java

```diff
@@ -170,72 +127,122 @@ public CurrentAssignmentBuilder withOwnedTopicPartitions(
+                // until the member has revoked the necessary partitions. They are
+                // considered revoked when they are not anymore reported in the
+                // owned partitions set in the ConsumerGroupHeartbeat API.
 
-            // Nothing to do.
-            case STABLE:
-                return member;
+                // If the member does not provide its owned partitions. We cannot
+                // progress.
+                if (ownedTopicPartitions == null) {
+                    return member;
+                }
+
+                // If the member provides its owned partitions. We verify if it still
+                // owns any of the revoked partitions. If it does, we cannot progress.
+                for (ConsumerGroupHeartbeatRequestData.TopicPartitions topicPartitions : ownedTopicPartitions) {
+                    for (Integer partitionId : topicPartitions.partitions()) {
+                        boolean stillHasRevokedPartition = member
+                            .partitionsPendingRevocation()
+                            .getOrDefault(topicPartitions.topicId(), Collections.emptySet())
+                            .contains(partitionId);
+                        if (stillHasRevokedPartition) {
+                            return member;
+                        }
+                    }
+                }
+
+                // When the member has revoked all the pending partitions, it can
+                // transition to the next epoch (current + 1) and we can reconcile
+                // its state towards the latest target assignment.
+                return computeNextAssignment(
+                    member.memberEpoch() + 1,
+                    member.assignedPartitions()
+                );
+
+            case UNRELEASED_PARTITIONS:
+                // When the member is in the UNRELEASED_PARTITIONS, we reconcile the
+                // member towards the latest target assignment. This will assign any
+                // of the unreleased partitions when they become available.
+                return computeNextAssignment(
+                    member.memberEpoch(),
+                    member.assignedPartitions()
+                );
+
+            case UNKNOWN:
+                // We could only end up in this state if a new state is added in the
+                // future and the group coordinator is downgraded. In this case, the
+                // best option is to fence the member to force it to rejoin the group
+                // without any partitions and to reconcile it again from scratch.
+                if (ownedTopicPartitions == null || !ownedTopicPartitions.isEmpty()) {
+                    throw new FencedMemberEpochException("The consumer group member is in a unknown state. "
+                        + "The member must abandon all its partitions and rejoin.");
+                }
+
+                return computeNextAssignment(
```

Comment:
Is the idea we only hit this case below on restart? Ie, we fence and force the member out of the group, but it will still be unknown on rejoining (just with no owned partitions)

Reply:
We would only hit this case if we ever introduce a new state in the future. I was trying to ensure that we would handle it somehow. My proposal here is to return the fenced member epoch error in order to force the member to abandon its partitions and rejoins the group.